entry_proxy/proxy_test.go

```diff
@@ -0,0 +1,213 @@
+				return
+			} else {
+				continue
+			}
+		}
+
+		l.conns = append(l.conns, conn)
+		go l.handleConnection(conn, len(l.conns)-1)
+	}
+}
+
+func (l *MortalService) handleConnection(conn net.Conn, id int) error {
+	defer func() {
+		log.Debugf("Closing connection #%d", id)
+		conn.Close()
+		l.conns[id] = nil
+	}()
+
+	log.Debugf("Starting connection #%d", id)
+	if err := l.connectionCallback(conn); err != nil {
+		return err
+	}
+	return nil
+}
+
+type AccumulatingListener struct {
+	net, address    string
+	buffer          bytes.Buffer
+	mortalService   *MortalService
+	hasProtocolInfo bool
+	hasAuthenticate bool
+	Received        chan bool
+}
+
+func NewAccumulatingListener(net, address string) *AccumulatingListener {
+	l := AccumulatingListener{
+		net:             net,
+		address:         address,
+		hasProtocolInfo: true,
+		hasAuthenticate: true,
+		Received:        make(chan bool, 0),
+	}
+	return &l
+}
+
+func (a *AccumulatingListener) Start() {
+	a.mortalService = NewMortalService(a.net, a.address, a.SessionWorker)
+	err := a.mortalService.Start()
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (a *AccumulatingListener) Stop() {
+	fmt.Println("AccumulatingListener STOP")
+	a.mortalService.Stop()
+}
+
+func (a *AccumulatingListener) SessionWorker(conn net.Conn) error {
+	connReader := bufio.NewReader(conn)
+	for {
+
+		line, err := connReader.ReadBytes('\n')
+		if err != nil {
+			//fmt.Println("AccumulatingListener read error:", err)
+		}
+		lineStr := strings.TrimSpace(string(line))
+		a.buffer.WriteString(lineStr + "\n")
+		a.Received <- true
+	}
+	return nil
+}
+
+type MockSNIParser struct{}
+
+func (m MockSNIParser) ServerNameFromConn(clientConn net.Conn) (string, net.Conn, error) {
+	return "Horse25519", clientConn, nil
+}
+
+type MockOnionResolver struct{}
+
+func (o *MockOnionResolver) ResolveToOnion(hostname string) (string, error) {
+	return "MockOnionHaHa", nil
+}
+
+type MockProxyDialer struct {
+	proxyNet  string
+	proxyAddr string
+}
+
+func NewMockProxyDialer(proxyNet, proxyAddr string) *MockProxyDialer {
+	d := MockProxyDialer{
+		proxyNet:  proxyNet,
+		proxyAddr: proxyAddr,
+	}
+	return &d
+}
+
+func (d *MockProxyDialer) Dial(targetServer string) (net.Conn, error) {
+	conn, err := net.Dial(d.proxyNet, d.proxyAddr)
+	return conn, err
+}
+
+func TestTLSProxy(t *testing.T) {
+	fakeTorNet := "tcp"
+	fakeTorAddr := "127.0.0.1:34628"
+	fakeTorListener := NewAccumulatingListener(fakeTorNet, fakeTorAddr)
+	fakeTorListener.Start()
+
+	proxyNet := "tcp"
+	proxyAddr := "127.0.0.1:44333"
+	proxy := NewTLSProxy(443, fakeTorNet, fakeTorAddr)
+	proxy.sniParser = MockSNIParser{}
+	proxy.resolver = &MockOnionResolver{}
+	proxy.dialer = NewMockProxyDialer(fakeTorNet, fakeTorAddr)
+	go proxy.Start(proxyNet, proxyAddr)
+
+	conn, err := net.Dial(proxyNet, proxyAddr)
+	if err != nil {
+		t.Errorf("failed to connect to proxy: %s", err)
+		t.Fail()
+	}
+
+	fmt.Fprint(conn, "meow\r\n")
+
+	<-fakeTorListener.Received
+
+	want := "meow\n"
```

Comment:
Can you create this variable before `fmt.Fprint(conn, "meow\r\n")` and reuse its value in it to prevent copy-paste?

Why these strings are not equal (`meow\r\n` and `meow\n`)?